core/src/main/java/org/jruby/RubyIO.java

```diff
@@ -1480,6 +1519,60 @@ public IRubyObject write(ThreadContext context, IRubyObject str, boolean nosync)
         return RubyFixnum.newFixnum(runtime, n);
     }
 
+    /**
+     * Write the given range of bytes to this IO.
+     *
+     * Equivalent to io_write_m with source bytes and no digging out any wrapped write IO (bytes will be written to this
+     * IO).
+     *
+     * @param context the current context
+     * @param bytes the bytes to write
+     * @param start start offset for writing
+     * @param length length to write
+     * @param encoding encoding of the bytes (will not be verified)
+     * @return the count of bytes written
+     */
+    public int write(ThreadContext context,  byte[] bytes, int start, int length, Encoding encoding) {
```

Comment:
do we really want context param passed around like for Ruby method impls?
we only need the runtime and from a Java ext perspective it would be nicer not to worry about context.
we could pass the runtime but I would love to also see a Java API like:
`public int write(byte[] bytes, int start, int length, Encoding encoding)`

Reply:
Passing Ruby means anything which involves a call or a callsite (and this is just the ones I can think of readily) will have to getCurrentContext which would be great to never do.  Maybe still do it but only in places where we don't care about performance.  Obviously, the other case, is any TC state we want directly.

Historically we started with Runtime and it wasn't until later we realize TC made much more sense as our global handle.  I would like to talk to you about how we can change ext to TC is the default.   My actual goal is not about TC at all but to normalize and formalize most of the common accesses of a JRuby ext to static methods which can then be a dictionary of mappings to MRI cext names (or not when that doesn't make sense).

Reply:
In this case the context ends up getting passed around mostly for access to the runtime, EncodingService, and metaclasses for raising errors at appropriate times, but the unavoidable use of context is to do an interruptible non-blocking write operation.

Any irreducible use of context means it should be passed in, since even after lots of work the context retrieval is a very heavy operation. See for example #5959 and https://github.com/eclipse/openj9/issues/7699.

This is not to say that I like passing it, because I totally agree it makes for a very cumbersome API, and the more we pass it, the more we avoid the threadlocal but the more we force people to pass it. I do agree that having non-context versions of these write (and eventually read) methods will be useful, and I will add them... but with lots of doco about the cost of ThreadContext retrieval and recommendations to acquire it once for hot read/write loops.

Reply:
I just want to highlight another problem with using Ruby over ThreadContext since the three of us have written stuff.  I think it goes beyond whether you need it or not.  You may eventually need it.  For example, we end up adding a dyncall when originally we had not planned on having a fall back to a dyn call.  Whoops.  At this point we:

  1. use getCurrentContext()
  2. OR we deprecate current method and make a new one
  
So in or first outcome we are stuck with this extra overhead (admittedly sometimes it doesn't matter).  In the second, we end up with a method which we will keep for 5-10 years.  We have many deprecated methods we have never even removed.

I think we need to just accept TC needs to be passed by default and be very particular about when we don't.  So I think we need to find a way of liking passing TC around.